assets/src/edit-story/components/canvas/useInsertElement.js

```diff
@@ -91,23 +90,96 @@ function useInsertElement() {
  * @param {?Object} props.mask The element's mask.
  * @return {!Object} The new element.
  */
-function createElement(type, { width, height, mask, ...rest }) {
-  const { isMaskable } = getDefinitionForType(type);
+function createElementForCanvas(
+  type,
+  { resource, x, y, width, height, rotationAngle, mask, ...rest }
+) {
+  const {
+    defaultAttributes,
+    isMaskable,
+    updateForResizeEvent,
+  } = getDefinitionForType(type);
+
+  const attrs = { ...defaultAttributes, ...rest };
+
+  // Width and height defaults. Width takes precedence.
+  const ratio =
+    resource && isNum(resource.width) && isNum(resource.height)
+      ? resource.width / resource.height
+      : 1;
+  if (!isNum(width)) {
+    if (isNum(height)) {
+      // Height is known: use aspect ratio.
+      width = height * ratio;
+    } else if (resource) {
+      // Resource is available: take resource's width with DPR, but limit
+      // to fit on the page.
+      width = Math.min(resource.width * DEFAULT_DPR, PAGE_WIDTH * 0.8);
+    } else {
+      // Default to half of page.
+      width = PAGE_WIDTH / 2;
+    }
+  }
+  if (!isNum(height) && updateForResizeEvent) {
+    // Try resize API with width-only direction.
+    height = updateForResizeEvent(attrs, RESIZE_WIDTH_DIRECTION, width).height;
+  }
+  if (!isNum(height)) {
```

Comment:
Doesn't this always overwrite the previously set `height = updateForResizeEvent(attrs, RESIZE_WIDTH_DIRECTION, width).height;` within `!isNum(height) && updateForResizeEvent`?

Reply:
Hmm. I think it shouldn't. Only if `updateForResizeEvent` is not specified or doesn't return a valid height.

Reply:
Ah, right, and then it should overwrite it anyway. 👍